### PR TITLE
Compile with `strictNullChecks`/`strictPropertyInitialization` (focus on calendar.ts and ecmascript.ts)

### PR DESCRIPTION
**Pull request by justingrant**
This PR focuses on unblocking `strictNullChecks` and `strictPropertyInitialization` in ecmascript.ts and calendar.ts. It's intended as a complement to @jens-ox's work in #105 to remove `GetIntrinsic`.

This PR:
* Enables `strictNullChecks: true` and `strictPropertyInitialization: true` in tsconfig.json.
* Improves calendar.ts TS types (this was most of the work in this PR)
* Removes all unnecessary use of `any` throughout the polyfill
* Updates types of ecmascript.ts and fixes a number of type bugs
* Refactors a few types in intl.ts
* A handful of trivial type changes in other files (ecmascript.ts and calendar.ts were 95%+ of the work).

There is a one-line runtime change (by itself in commit https://github.com/js-temporal/temporal-polyfill/pull/109/commits/fac8f4a2d839be420f4938184099050d79641fcb) but otherwise there should be no runtime-visible changes in this PR, only TS type changes. This means that, aside from that one line in calendar.ts, the transpiled output should be the same before and after this PR, except for parentheses used  for type casts. 

Other runtime changes were split out into other PRs, all of which have already been merged except #112 which will hopefully land soon. This PR is rebased on #112.

### CODE REVIEW

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:3437]: I'm not very familiar with TS best practices, but should this be something like `let relativeTo: Temporal.PlainDate? = undefined;`?

**Review by ptomato — COMMENTED** [review]
I didn't go through calendar.mjs and ecmascript.mjs too carefully, as this is a draft, but let me know what kind of reviews would be helpful! FWIW, it would be easier to review if you could split it into a series of commits that each fix one type of bug (although I realize that may not be possible; when you fix the types on one thing, maybe it "cascades" into another thing?)

**Review by justingrant — COMMENTED** [review]
> I didn't go through calendar.mjs and ecmascript.mjs too carefully, as this is a draft, but let me know what kind of reviews would be helpful! 

@ptomato Lol, I actually didn't expect any reviews at all until this was further along, so your initial notes are more than I expected! I'll ping when it's ready for a more thorough review.

The best assistance at this point would be to help review and land the proposal-temporal PRs listed in the OP of this PR, because those all came from issues I discovered here while making these TS changes.  The sooner those PRs are merged, the sooner this PR can get wrapped up.

> FWIW, it would be easier to review if you could split it into a series of commits that each fix one type of bug (although I realize that may not be possible; when you fix the types on one thing, maybe it "cascades" into another thing?)

My current plan is to merge all the related proposal-temporal PRs, then port those PRs here, and then rebase this PR on top which should reduce its scope a lot. Furthermore, be removing those PRs it should limit this PR to almost entirely type changes: either none or only a few runtime-visible changes and zero changes that affect tests. It should be both easier and safer to review a big PR with only type changes vs. a mix of type a runtime changes. At least that's the goal. 😄 

But you're right-- it's hard to make these kinds of type changes piecemeal, because once you pull on a thread of types it's hard to split things out.  Luckily, once the runtime changes are removed it should be a lot clearer what's different. 

Sound like an OK plan? Holler with any suggestions to make it easier.

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts]: Update: here's a comment from the commit I just pushed that includes the new names of the types and an explanation for each: 

```
// Some ES implementations don't include ECMA 402. For this reason, it's helpful
// to ensure a clean separation between the ISO calendar implementation which is
// a part of ECMA 262 and the non-ISO calendar implementation which requires
// ECMA 402.
//
// To ensure this separation, the implementation is split. The `NonIsoImpl`
// interface is the top-level implementation for all non-ISO calendars. This
// type has the same shape as the ECMA 262-only ISO calendar implementation so
// can use the same callers, tests, etc.
//
// A derived interface `NonIsoImplWithHelper` adds a `helper` property that
// includes the remaining non-ISO implementation properties and methods beyond
// the ISO implementation above. The `helper` property's shape is a base
// singleton object common to all calendars (`HelperSharedImpl`) that's extended
// (interface `HelperPerCalendarImpl`) with implementation that varies for each
// calendar.
```

File: lib/ecmascript.ts
Comment (ptomato) [reply lib/ecmascript.ts:455]: Sounds good, I think that's an equally valid solution to the problem.

File: lib/ecmascript.ts
Comment (justingrant) [reply lib/ecmascript.ts:1065]: Update: we removed it from proposal-temporal too. https://github.com/tc39/proposal-temporal/pull/1980

**Comment by justingrant** [comment]
I split this PR up into a bunch of commits to make reviewing easier, esp. for calendar.ts.

It's currently rebased on top of #112 which will hopefully land soon.

**Comment by justingrant** [comment]
Oh, and it's ready for review now. Reviews may not happen until after the new year.

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts]: Do any of the overriding calendars use the month and day to determine `inLeapYear`? If not, maybe we could just pass `year: number` to this method instead of what is effectively `calendarDate: { year: number }`?

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:510]: No idea if this would work, but is it possible to do something like `mergeFields: typeof Temporal.Calendar.prototype.mergeFields`? Just out of curiosity.

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:2139]: Also not sure if this is valid TypeScript but can we do something like `interface ChineseMonthInfo = Required<ChineseDraftMonthInfo>`?

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:1445]: Wouldn't this make TypeScript think that the check below (`if (year === undefined ...)`) was redundant? Should the type here be unioned with `{ year?: number, day?: number }` instead?

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:1462]: Does the `as number` work because the `inLeapYear` result makes sure that the `undefined` variant is never selected?

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:1540]: Is it worth defining a type alias for this, or is it just going to go away anyway when we use ES6 classes here?

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:2167]: I don't quite get this one, why do we need `as unknown` now if it worked before? (Does this mean we should actually be typing `temporalTimeZoneLike` as something like `Temporal.TimeZoneProtocol | { timeZone: Temporal.TimeZoneProtocol }`?)

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:4760]: I'm not a good judge of what's idiomatic, but might it be better to type this as `JSBI | undefined` and use `!` in the branch where we know it is not undefined?

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:5135]: I'm not familiar enough with TypeScript to know what these declarations do if they don't have function bodies. Do they just serve as alternate type signatures for the function below that does have a body?

File: lib/intl.ts
Comment (ptomato) [inline lib/intl.ts:19]: Should this be `import type`?

File: lib/intrinsicclass.ts
Comment (ptomato) [inline lib/intrinsicclass.ts:99]: Out of curiosity, why doesn't `Object.getOwnPropertyDescriptor` have type `PropertyDescriptor` already?

File: lib/plaintime.ts
Comment (ptomato) [inline lib/plaintime.ts:46]: Could be `options?: TemporalTimeToStringOptions`?

**Review by ptomato — APPROVED** [review]
Great stuff, I think you'll find that my comments reflect my ignorance of TypeScript idioms rather than problems with the code. Thanks, by the way, for the separate commits and the explanations in the commit messages, that really made reviewing this a breeze.

**Review by justingrant — COMMENTED** [review]
Thanks so much @ptomato for the thorough review!  There were two things you found that should result in code changes, but both of them are wider in scope than this PR, so I split them into separate issues. 

For your other comments, I don't think changes are needed, but if you disagree please let me know!  Otherwise, I think the PR code is still good to merge unless I hear differently from @12wrigja. I'll wait a few days for James to catch up before merging.

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:4760]: I think I'd type this as `dayLengthNs: JSBI|undefined` and then cast using `!` along with a comment explaining why we know for certain it's been assigned a value there. This might force us to write this out multiple times, but I think that it's best to document these assumptions.

File: lib/intrinsicclass.ts
Comment (12wrigja) [reply lib/intrinsicclass.ts:99]: This is probably somewhere that I'd use `!` instead of `as`, but it sounds like there's an ESLint rule against that?

File: lib/timezone.ts
Comment (12wrigja) [inline lib/timezone.ts:128]: I'm guessing GetSlot is defined to return `JSBI` for this key, so why doesn't this slot always contain this value? Should this be changed in slots to return `JSBI|undefined`?

**Review by 12wrigja — APPROVED** [review]

File: lib/calendar.ts
Comment (ptomato) [reply lib/calendar.ts:510]: Is this something that will go away when we use ES6 classes here? I ask because I find these types a bit difficult to read...

File: lib/plaintime.ts
Comment (ptomato) [reply lib/plaintime.ts:46]: That playground link won't load for me, unfortunately (it gets stuck on "Downloading TypeScript"...) but are you referring to the function length?

Comment (justingrant) [reply lib/plaintime.ts:46]: Oh, weird. Never heard of that TS Playground symptom. Anyway, here's what's in there: 

```ts
// TS code:
function f1(x: string | undefined = undefined) {}
function f2(x?: string) {}

// JS code that's emitted by the transpiler
function f1(x = undefined) { }
function f2(x) { }
```

The `= undefined` is different between the two, so I assumed it was better to match the runtime output whenever possible.

Comment (ptomato) [reply lib/plaintime.ts:46]: Probably inconsequential because the length of that function is not observable, but I can see that divergences might make it harder to debug.

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts:510]: Here's what the equivalent will look like in ES6 classes: https://github.com/js-temporal/temporal-polyfill/pull/113/files#diff-0751307e1b8622b923dfb7f18f7aed915f397cd1f6a67a5465d9672db48839f5R478

Is that better?

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts:1380]: 👍  I did this in #113, so I'll plan to fix it there instead of here.

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts:492]: Yep!  See #113.

**Review by justingrant — COMMENTED** [review]
I just pushed changes to resolve open issues in code review comments, so unless there are additional issues I'll plan to merge this and then finalize #113 to convert calendar.ts to ES6.

File: lib/calendar.ts
Comment (ptomato) [reply lib/calendar.ts:510]: I have a vague hunch that this is more complicated than it needs to be, but I'm not prepared to spend time to prove it, so I won't complain anymore :smile:

Comment (justingrant) [reply lib/calendar.ts:510]: "I have a vague hunch that this is more complicated than it needs to be" could be the motto of half of my time on this planet, I think. ;-)

File: lib/ecmascript.ts
Comment (justingrant) [reply lib/ecmascript.ts]: OK, I checked in @12wrigja's suggestion.

File: lib/intrinsicclass.ts
Comment (justingrant) [reply lib/intrinsicclass.ts]: Yeah. I ended up checking in a change here that's the same as the other case you commented about: turn off the ESLint rule for this line (and a few lines below in a similar loop) and add a comment explaining why.

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts]: This type got renamed in a later commit. If you're still concerned about the interface names, let's discuss over in #113 which has a lot more renaming!

File: lib/timezone.ts
Comment (12wrigja) [reply lib/timezone.ts:128]: Ah gotcha - yeah I missed that it was reassigned later. I agree that this solution is fine, but certainly far less than ideal.

**Pull request merged**
